### PR TITLE
Fix access error when accessing product creation modal

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.1.x
| Description?      | https://github.com/PrestaShop/PrestaShop/issues/30202. Instead of showing error, we are not showing the link at all if user has no access to it. In case its a direct link, then it will redirect back to default page (that is set in employee edit page) and show the access denied error
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      |  https://github.com/PrestaShop/PrestaShop/issues/30202 Instead of showing error, we are not showing the link at all if user has no access to it.
| Fixed ticket?     | https://github.com/PrestaShop/PrestaShop/issues/30202 
| Related PRs       | 
| Sponsor company   |

### CODE REVIEW

**Comment by zuk3975** [comment]
blocked. Waiting for PM approval.
pint @MatShir Is it ok if I will not show the create product button at all when user has no rights to access the link?
By default legacy did show them, but once you click you get redirected with error "access denied".
Now I can either make the modal show the access denied error (which from technical perspective only possible by a bit "shitty" workaround), or I can do it technically "cleaner" and not show the link at all.

Depending on your answer I'll have to leave one or another commit

**Comment by MatShir** [comment]
I like the idea:
Tell me what you can do but not show the link is good for me. I'm thinking that disabling the input could work as well, seems a bit better if you can. 
Let me know and I will update the specs accordingly

**Review by kpodemski — DISMISSED** [review]

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php
Comment (FabienPapet) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]: With proper import

```suggestion
        if (!$this->isGranted(Permission::CREATE, $securitySubject)) {
```

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php
Comment (FabienPapet) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php:140]: Maybe add the permission needed as a parameter as well

Comment (FabienPapet) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php:140]: Like so ?

```suggestion
            'layoutHeaderToolbarBtn' => $this->getProductToolbarButtons(Permission::CREATE, $request->get('_legacy_controller')),
```

Comment (zuk3975) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php:140]: every button in toolbar may need different permission, so no. (even though now there is only one button, but anyway, it wouldn't be right to pass one permission)

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php
Comment (zuk3975) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php]: there is no such class, but good idea to use const. There is PageVoter::CREATE

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php
Comment (FabienPapet) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php:140]: OK

**Review by jolelievre — APPROVED** [review]
Thanks @zuk3975 

seems like an appropriate solution, however it will only be temporary for quick access link, we can check the permission now because we hard-coded a special case for this link but it wouldn't work for any other link and we can't add specific rules for each link.

It's enough for now but once we clear the quick access in v9 once the new product page is officially released we'll have to think about another approach. By then maybe the menu/header will have been refactored and we'll be able to have a better solution.

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php
Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php:140]: Better check the permission in the method itself, for now, there's only one button but if others are added they may be linked to another permission, the controller associated with this page however should not change.

**Review by matks — APPROVED** [review]

**Review by florine2623 — COMMENTED** [review]
Hello @zuk3975 ,

Link to automated tests : https://github.com/florine2623/testing_pr/actions/runs/4261000997

**Case 1:** ✅ 

> If the user doesn't have the right to edit or add product
> Then the link to the quick access is hidden

Screenshot 2023-02-24 at 15 47 29

**Case 2:**

> If the user doesn't have the right to edit or add product
> Then the button add/edit is hidden

Screenshot 2023-02-24 at 16 03 34

:x: BUT, the button edit (pen icon) is still displayed, what is the exact wanted behavior ?

**Case 3:** ✅ 

> If the user accesses to a direct link of add or edit product
> Then he is redirected to the default page

https://user-images.githubusercontent.com/16019289/221212937-e5f32097-d9f9-44dd-8403-4a8148015e22.mov

**Comment by zuk3975** [comment]
Thanks @florine2623,
| BUT, the button edit (pen icon) is still displayed, what is the exact wanted behavior ?

It is expected behavior for now. In future we will probably want to hide actions in grid too, but for now its too much work. Edit action is not in modal, so it should just redirect back to default with flash message and not cause any trouble.

**Comment by jolelievre** [comment]
@florine2623 I agree about the edit buttons, it's not the scope of this PR and it's not handled in all the other grids as far as I know. It should be handled in a dedicated Epic probably if we want to generalize this behaviour to all grids

So if it's the only "bug" I'd say this PR is fine for the scope it focuses on

**Review by florine2623 — APPROVED** [review]
Hello @zuk3975 and @jolelievre ,

Thanks for the details.
That was the only thing I found, so it is QA ✅ !

Thanks!

**Pull request merged**
